src/Functions/URL/resolveRelativeURL.cpp

```diff
@@ -0,0 +1,218 @@
+    if (has_subsequent_slash)
+    {
+        return pos+3;
+    }
+    return url_end;
+}
+
+Pos getDomainNameEndPos(const char * url_begin, const char * url_end)
+{
+    Pos pos = find_first_symbols<':'>(url_begin, url_end);
+    if (pos == url_end)
+    {
+        return pos;
+    }
+    bool has_subsequent_slash = pos[1] == '/' && pos[2] == '/';
+    if (has_subsequent_slash)
+    {
+        pos = find_first_symbols<'/'>(pos + 3, url_end);
+        return pos;
+    }
+    return url_end;
+}
+
+class FunctionrResolveRelativeURL : public IFunction
+{
+public:
+    static constexpr auto name = "resolveRelativeURL";
+    static FunctionPtr create(ContextPtr) { return std::make_shared<FunctionrResolveRelativeURL>(); }
+
+    String getName() const override { return name; }
+
+    size_t getNumberOfArguments() const override { return 2; }
+
+    bool useDefaultImplementationForConstants() const override { return true; }
+    ColumnNumbers getArgumentsThatAreAlwaysConstant() const override { return {1}; }
+
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        if (!isString(arguments[0]))
+            throw Exception(
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type {} of argument of function {}", arguments[0]->getName(), getName());
+
+        if (!isString(arguments[1]) && !isArray(arguments[1]))
+            throw Exception(
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type {} of argument of function {}", arguments[1]->getName(), getName());
+
+        return std::make_shared<DataTypeString>();
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t /*input_rows_count*/) const override
+    {
+        const ColumnPtr relative_url_column = arguments[0].column;
+        const ColumnPtr base_url_column = arguments[1].column;
+
+        const ColumnString * relative_url = checkAndGetColumn<ColumnString>(relative_url_column.get());
+
+        const auto * base_url = checkAndGetColumnConstData<ColumnString>(base_url_column.get());
+
+        if (!relative_url)
+        {
+            throw Exception(
+                ErrorCodes::ILLEGAL_COLUMN, "Illegal column {} of argument of function {}", arguments[0].column->getName(), getName());
+        }
+
+        if (!base_url)
+        {
+            throw Exception(
+                ErrorCodes::ILLEGAL_COLUMN, "Illegal column {} of argument of function {}", arguments[1].column->getName(), getName());
+        }
+
+        auto col_res = ColumnString::create();
+
+        ColumnString::Chars & vec_res = col_res->getChars();
+        ColumnString::Offsets & offsets_res = col_res->getOffsets();
+        vector(relative_url->getChars(), relative_url->getOffsets(), base_url->getChars(), base_url->getOffsets(), vec_res, offsets_res);
```

Comment:
Since we have `getArgumentsThatAreAlwaysConstant` we may want to have vectorConstant variant. But it will require to disable `useDefaultImplementationForConstants` and handle all cases manually, so it's ok to have only a bit sub-optimal generic `vector` variant for simplicity

Reply:
Sorry, I can not understand what you wanted to tell me, can you please explain more clearly?

Reply:
Some functions have different implementations for constant and vector arguments. For example `func(s, 'aaa')` and `func(s1, s2)` can be executed in a different way, since in the first case for each `s` the value of the second would be `'aaa'`. In that case we can read its value once and do not access column for second argument on each iteration. For example:

https://github.com/ClickHouse/ClickHouse/blob/4d27f87be89fa2c3eeb914398334eb9dac0cd32f/src/Functions/FunctionsStringSearch.h#L185-L205

But it makes code more complicated, so let's have it like it's done now (and `extractURLParameter` does the same)